File: frontend/src/app/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Comment } from '../models';

@Injectable({ providedIn: 'root' })
export class CommentService {
  private api = `${environment.apiUrl}/comments`;

  constructor(private http: HttpClient) {}

  addComment(comment: Partial<Comment>): Observable<Comment> {
    return this.http.post<Comment>(this.api, comment);
  }

  getByFile(fileId: string): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${this.api}/file/${fileId}`);
  }

  getByProject(projectId: string): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${this.api}/project/${projectId}`);
  }

  getReplies(commentId: string): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${this.api}/${commentId}/replies`);
  }

  getByLine(fileId: string, lineNumber: number): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${this.api}/file/${fileId}/line/${lineNumber}`);
  }

  updateComment(commentId: string, content: string): Observable<Comment> {
    return this.http.put<Comment>(`${this.api}/${commentId}`, { content });
  }

  deleteComment(commentId: string): Observable<any> {
    return this.http.delete(`${this.api}/${commentId}`);
  }

  resolveComment(commentId: string): Observable<Comment> {
    return this.http.put<Comment>(`${this.api}/${commentId}/resolve`, {});
  }

  unresolveComment(commentId: string): Observable<Comment> {
    return this.http.put<Comment>(`${this.api}/${commentId}/unresolve`, {});
  }

  getCount(fileId: string): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.api}/file/${fileId}/count`);
  }
}

File: backend/src/main/java/com/codesync/comment/service/CommentService.java
package com.codesync.comment.service;

import com.codesync.comment.entity.Comment;
import java.util.List;

public interface CommentService {
    Comment addComment(Comment comment);
    Comment getCommentById(String commentId);
    List<Comment> getByFile(String fileId);
    List<Comment> getByProject(String projectId);
    List<Comment> getReplies(String parentCommentId);
    List<Comment> getByLine(String fileId, int lineNumber);
    Comment updateComment(String commentId, String newContent);
    void deleteComment(String commentId);
    Comment resolveComment(String commentId);
    Comment unresolveComment(String commentId);
    long getCommentCount(String fileId);
}
